### ISSUE TITLE
TODOs for IRB support

### ISSUE THREAD

**Issue by st0012** [ruby/yarp#930]
Following @jemmaissroff's suggestion, here's a list of things IRB'd need in order to use YARP instead of Ripper:

- A way to simulate the output of `Ripper.sexp`. Maybe something like `YARP.lex(code).value.to_array`?
    - https://github.com/ruby/irb/blob/master/lib/irb.rb#L889
    - https://github.com/ruby/irb/blob/df6907aca9c618561537e37da0ef0626f675c42c/lib/irb/cmd/nop.rb#L28
- A way to do simulate `Ripper::Lexer#scan`. I believe `YARP.lex_compat` is created for this.
    - However, the result needs to have `:on_sp` token as noted in #722
    - https://github.com/ruby/irb/blob/df6907aca9c618561537e37da0ef0626f675c42c/lib/irb/color.rb#L200
    - https://github.com/ruby/irb/blob/df6907aca9c618561537e37da0ef0626f675c42c/lib/irb/ruby-lex.rb#L164
- IRB also needs a way to map Ripper's state bits, like `Ripper::EXPR_CMDARG` with YARP's, and access them through similar constants, like `YARP::EXPR_CMDARG`?
    - Perhaps IRB could first simplify/change its state checks to make this easier? ([example](https://github.com/ruby/irb/blob/df6907aca9c618561537e37da0ef0626f675c42c/lib/irb/ruby-lex.rb#L358-L370)). @tompng is it possible we reduce `RubyLex `'s coupling on Ripper state bits?
- Finally, IRB will need a way to know if the current parser is YARP so it can switch to use different APIs/constants.

**Comment by tompng** [comment]
> is it possible we reduce RubyLex's coupling on Ripper state bits?

Yes, it's possible. For example in https://github.com/ruby/irb/pull/500, use of `token.state.allbits?(Ripper::EXPR_SOMETHING)` is reduced. (17 deletion, 4 addition)

**Comment by kddnewton** [comment]
Hey @st0012 - I know it's been half a year, but I just wanted to check in on where we are with this. I know a lot of changes have gone into IRB since you opened this issue. Is everything here still relevant/necessary?
